{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Enter start codon: atg\n",
      "Good Job! That's the correct start codon\n",
      "Enter a DNA triplet or stop codon (up to 9 entries!): ttt\n",
      "['Met', 'Phe']\n",
      "Enter a DNA triplet or stop codon (up to 9 entries!): aaa\n",
      "['Met', 'Phe', 'Lys']\n",
      "Enter a DNA triplet or stop codon (up to 9 entries!): cag\n",
      "['Met', 'Phe', 'Lys', 'Gln']\n",
      "Enter a DNA triplet or stop codon (up to 9 entries!): gag\n",
      "['Met', 'Phe', 'Lys', 'Gln', 'Glu']\n",
      "Enter a DNA triplet or stop codon (up to 9 entries!): tag\n"
     ]
    }
   ],
   "source": [
    "import graphics\n",
    "# Nucleotide List\n",
    "DNA = [\"A\", \"C\", \"G\", \"T\"]\n",
    "\n",
    "codons = { \n",
    "    'GCT' : 'Ala', 'GCC' : 'Ala', 'GCA' : 'Ala', 'GCG' : 'Ala',\n",
    "    'TGT' : 'Cys', 'TGC' : 'Cys',\n",
    "    'GAT' : 'Asp', 'GAC' : 'Asp',\n",
    "    'GAA' : 'Glu', 'GAG' : 'Glu',\n",
    "    'TTT' : 'Phe', 'TTC' : 'Phe',\n",
    "    'GGT' : 'Gly', 'GGC' : 'Gly', 'GGA' : 'Gly', 'GGG' : 'Gly',\n",
    "    'CAT' : 'His', 'CAC' : 'His',\n",
    "    'ATT' : 'Ile', 'ATC' : 'Ile', 'ATA' : 'Ile',\n",
    "    'AAA' : 'Lys', 'AAG' : 'Lys',\n",
    "    'CTT' : 'Leu', 'CTC' : 'Leu', 'CTA' : 'Leu', 'CTG' : 'Leu', 'TTA' : 'Leu', 'TTG' : 'Leu',\n",
    "    'ATG' : 'Met',\n",
    "    'AAT' : 'Asn', 'AAC' : 'Asn',\n",
    "    'CCT' : 'Pro', 'CCC' : 'Pro', 'CCA' : 'Pro', 'CCG' : 'Pro',\n",
    "    'CAA' : 'Gln', 'CAG' : 'Gln',\n",
    "    'CGT' : 'Arg', 'CGC' : 'Arg', 'CGA' : 'Arg', 'CGG' : 'Arg', 'AGA' : 'Arg', 'AGG' : 'Arg',\n",
    "    'TCT' : 'Ser', 'TCC' : 'Ser', 'TCA' : 'Ser', 'TCG' : 'Ser', 'AGT' : 'Ser', 'AGC' : 'Ser',\n",
    "    'ACT' : 'Thr', 'ACC' : 'Thr', 'ACA' : 'Thr',\n",
    "    'ACG' : 'Thr',\n",
    "    'GTT' : 'Val', 'GTC' : 'Val', 'GTA' : 'Val', 'GTG' : 'Val',\n",
    "    'TGG' : 'Trp',\n",
    "    'TAT' : 'Tyr', 'TAC' : 'Tyr'\n",
    "           }\n",
    "\n",
    "# Stop codon list\n",
    "stop = ('TAA', 'TAG', 'TGA', 'taa', 'tag', 'tga')\n",
    "\n",
    "# Codon Dictionary\n",
    "DNA_codon_list = ['Met']\n",
    "\n",
    "def validity_check(codon):\n",
    "    codon = codon.upper()\n",
    "    if set(codon).issubset(DNA):\n",
    "        return three_base_pairs(codon)\n",
    "    else:\n",
    "        print(\"This is an invalid DNA sequence\")\n",
    "\n",
    "def three_base_pairs(codon):\n",
    "    if len(codon)/3 == 1:\n",
    "        return DNA_list(codon)\n",
    "    else:\n",
    "        print(\"This is an invalid DNA codon. Must be three bases in length.\")\n",
    "        \n",
    "def DNA_list(codon):\n",
    "    global DNA_codon_list\n",
    "    DNA_codon_list.append(codons[codon])\n",
    "    print(DNA_codon_list)\n",
    "        \n",
    "def main():\n",
    "    codon = input(\"Enter start codon: \")\n",
    "    if codon == 'ATG' or codon == 'atg':\n",
    "        print(\"Good Job! That's the correct start codon\")\n",
    "    else:\n",
    "        print(\"Sorry, that's not correct. ATG is the start codon for DNA, you may continue anyways.\")\n",
    "    codon = input(\"Enter a DNA triplet or stop codon (up to 9 entries!): \")\n",
    "    while codon not in stop and len(DNA_codon_list) < 10:\n",
    "        validity_check(codon)\n",
    "        codon = input(\"Enter a DNA triplet or stop codon (up to 9 entries!): \")\n",
    "\n",
    "main()\n",
    "\n",
    "def draw():\n",
    "    category = { 'Ala': '#00AAA0', 'Val': '#00AAA0', 'Cys': '#00AAA0', 'Pro': '#00AAA0', 'Gly': '#00AAA0',\n",
    "                'Leu': '#00AAA0', 'Ile': '#00AAA0', 'Met': \"#00AAA0\", 'Trp': '#00AAA0', 'Phe':'#00AAA0', \n",
    "                'Ser': '#FF7A5A', 'Thr': '#FF7A5A', 'Tyr': '#FF7A5A', 'Asn': '#FF7A5A', 'Gln':'#FF7A5A',\n",
    "                'Lys': '#001D71', 'Arg': '#001D71', 'His':'#001D71',\n",
    "                'Asp': '#853C97', 'Glu':'#853C97'\n",
    "               }\n",
    "\n",
    "    win = graphics.GraphWin(\"Primary Structure of Protein\", 1000, 500)\n",
    "    for i, amino_acids in enumerate(DNA_codon_list):\n",
    "        pos = 75 + 90 * i\n",
    "        pt = graphics.Point(x = pos, y = 250)  \n",
    "        c = graphics.Circle(pt, 45)\n",
    "        c.draw(win)\n",
    "        c.setFill(category[amino_acids])\n",
    "        text_label = graphics.Text(graphics.Point(x = pos, y = 250), amino_acids)\n",
    "        text_label.draw(win)\n",
    "        text_label.setFill(\"white\")\n",
    "    text1 = graphics.Text(graphics.Point(500, 25), \"What a pretty protein!\")\n",
    "    text2 = graphics.Text(graphics.Point(500, 50), \"Teal = Non-Polar Amino Acids\")\n",
    "    text3 = graphics.Text(graphics.Point(500, 70), \"Coral = Polar Amino Acids\")\n",
    "    text4 = graphics.Text(graphics.Point(500, 90), \"Navy = Basic Amino Acids\")\n",
    "    text5 = graphics.Text(graphics.Point(500, 110), \"Purple = Acidic Amino Acids\")\n",
    "    text6 = graphics.Text(graphics.Point(500, 450), \"Click the screen to denature your protein!\")\n",
    "    text1.draw(win)\n",
    "    text2.draw(win)\n",
    "    text3.draw(win)\n",
    "    text4.draw(win)\n",
    "    text5.draw(win)\n",
    "    text6.draw(win)\n",
    "    win.getMouse()\n",
    "    win.close()\n",
    "draw()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0\n",
      "1\n",
      "2\n",
      "3\n",
      "4\n",
      "5\n",
      "6\n",
      "7\n",
      "8\n",
      "9\n",
      "10\n",
      "11\n",
      "12\n",
      "13\n",
      "14\n",
      "15\n",
      "16\n",
      "17\n",
      "18\n",
      "19\n",
      "20\n"
     ]
    }
   ],
   "source": [
    "def insulin():\n",
    "    for place, aa in enumerate(insulin_seq_a):\n",
    "        print(place)\n",
    "   # for aa in insulin_seq_b:\n",
    "#         if insulin_seq_b[aa] < 21:\n",
    "#             pass\n",
    "#         elif insulin_seq_b[aa] <= 24:\n",
    "#             print(aa)\n",
    "#         else:\n",
    "#             print(aa)\n",
    "\n",
    "\n",
    "insulin_seq_a = ['Gly', 'Ile', 'Val', 'Glu', 'Gln', 'Cys', 'Cys', 'Thr', 'Ser', 'Ile', 'Cys', 'Ser', 'Leu', 'Tyr', \n",
    "                 'Gln', 'Leu', 'Glu', 'Asn', 'Tyr', 'Cys', 'Gly']\n",
    "\n",
    "insulin_seq_b = ['Phe', 'Val', 'Asn', 'Gln', 'His', 'Leu', 'Cys', 'Gly', 'Ser', 'His', 'Leu', 'Val', 'Glu', 'Ala', \n",
    "                 'Leu', 'Tyr', 'Leu', 'Val', 'Cys', 'Gly', 'Glu', 'Arg', 'Gly', 'Phe', 'Phe', 'Tyr', 'Thr', 'Pro', \n",
    "                 'Lys', 'Thr', 'Arg', 'Arg']\n",
    "\n",
    "insulin()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "import graphics\n",
    "\n",
    "insulin_seq_a = ['Gly', 'Ile', 'Val', 'Glu', 'Gln', 'Cys', 'Cys', 'Thr', 'Ser', 'Ile', 'Cys', 'Ser', 'Leu', 'Tyr', \n",
    "                 'Gln', 'Leu', 'Glu', 'Asn', 'Tyr', 'Cys', 'Gly']\n",
    "\n",
    "insulin_seq_b = ['Phe', 'Val', 'Asn', 'Gln', 'His', 'Leu', 'Cys', 'Gly', 'Ser', 'His', 'Leu', 'Val', 'Glu', 'Ala', \n",
    "                 'Leu', 'Tyr', 'Leu', 'Val', 'Cys', 'Gly', 'Glu', 'Arg', 'Gly', 'Phe', 'Phe', 'Tyr', 'Thr', 'Pro', \n",
    "                 'Lys', 'Thr', 'Arg', 'Arg']\n",
    "\n",
    "def draw():\n",
    "    category = { 'Ala': '#00AAA0', 'Val': '#00AAA0', 'Cys': '#00AAA0', 'Pro': '#00AAA0', 'Gly': '#00AAA0',\n",
    "                'Leu': '#00AAA0', 'Ile': '#00AAA0', 'Met': \"#00AAA0\", 'Trp': '#00AAA0', 'Phe':'#00AAA0', \n",
    "                'Ser': '#FF7A5A', 'Thr': '#FF7A5A', 'Tyr': '#FF7A5A', 'Asn': '#FF7A5A', 'Gln':'#FF7A5A',\n",
    "                'Lys': '#001D71', 'Arg': '#001D71', 'His':'#001D71',\n",
    "                'Asp': '#853C97', 'Glu':'#853C97'\n",
    "               }\n",
    "\n",
    "    win = graphics.GraphWin(\"Primary Structure of Protein\", 1000, 500)\n",
    "    for i, amino_acids in enumerate(insulin_seq_a):\n",
    "        pos = 80 + 40 * i\n",
    "        pt = graphics.Point(x = pos, y = 100)  \n",
    "        c = graphics.Circle(pt, 20)\n",
    "        c.draw(win)\n",
    "        c.setFill(category[amino_acids])\n",
    "        text_label = graphics.Text(graphics.Point(x = pos, y = 100), amino_acids)\n",
    "        text_label.draw(win)\n",
    "        text_label.setFill(\"white\")\n",
    "    for i, amino_acids in enumerate(insulin_seq_b):\n",
    "        if i < 21:\n",
    "            pos = 80 + 40 * i\n",
    "            pt = graphics.Point(x = pos, y = 240)  \n",
    "            c = graphics.Circle(pt, 20)\n",
    "            c.draw(win)\n",
    "            c.setFill(category[amino_acids])\n",
    "            text_label = graphics.Text(graphics.Point(x = pos, y = 240), amino_acids)\n",
    "            text_label.draw(win)\n",
    "            text_label.setFill(\"white\")\n",
    "        elif i >= 21 and i < 25:\n",
    "            pos = 240 + 40 * (i-20)\n",
    "            pt = graphics.Point(x = 880, y = pos)  \n",
    "            c = graphics.Circle(pt, 20)\n",
    "            c.draw(win)\n",
    "            c.setFill(category[amino_acids])\n",
    "            text_label = graphics.Text(graphics.Point(x = 880, y = pos), amino_acids)\n",
    "            text_label.draw(win)\n",
    "            text_label.setFill(\"white\")\n",
    "        elif i >=25:\n",
    "            pos = 880 - 40 * (i-24)\n",
    "            pt = graphics.Point(x = pos, y = 400)  \n",
    "            c = graphics.Circle(pt, 20)\n",
    "            c.draw(win)\n",
    "            c.setFill(category[amino_acids])\n",
    "            text_label = graphics.Text(graphics.Point(x = pos, y = 400), amino_acids)\n",
    "            text_label.draw(win)\n",
    "            text_label.setFill(\"white\")\n",
    "    texta = graphics.Text(graphics.Point(100, 60), \"A Chain\")\n",
    "    textb = graphics.Text(graphics.Point(100, 290), \"B Chain\")\n",
    "    text_title = graphics.Text(graphics.Point(200, 400), \"A model of human insulin\")\n",
    "    text_title.setSize(24)\n",
    "    txt_s1l = graphics.Text(graphics.Point(280, 30), \"S\")\n",
    "    txt_s1r = graphics.Text(graphics.Point(480, 30), \"S\")\n",
    "    line1_h = graphics.Line(graphics.Point(290,30), graphics.Point(470,30))\n",
    "    line1_vl = graphics.Line(graphics.Point(280,40), graphics.Point(280,80))\n",
    "    line1_vr = graphics.Line(graphics.Point(480,40), graphics.Point(480,80))\n",
    "    txt_s2t = graphics.Text(graphics.Point(320, 155), \"S\")\n",
    "    txt_s2b = graphics.Text(graphics.Point(320, 185), \"S\")\n",
    "    line2_t = graphics.Line(graphics.Point(320,120), graphics.Point(320,145))\n",
    "    line2_m = graphics.Line(graphics.Point(320,165), graphics.Point(320,175))\n",
    "    line2_b = graphics.Line(graphics.Point(320,195), graphics.Point(320,220))\n",
    "    txt_s3t = graphics.Text(graphics.Point(820, 155), \"S\")\n",
    "    txt_s3b = graphics.Text(graphics.Point(820, 185), \"S\")\n",
    "    line3_t = graphics.Line(graphics.Point(820,145), graphics.Point(840,120))\n",
    "    line3_m = graphics.Line(graphics.Point(820,165), graphics.Point(820,175))\n",
    "    line3_b = graphics.Line(graphics.Point(800,220), graphics.Point(820,195))\n",
    "    texta.draw(win)\n",
    "    textb.draw(win)\n",
    "    text_title.draw(win)\n",
    "    txt_s1l.draw(win)\n",
    "    txt_s1r.draw(win)\n",
    "    txt_s2t.draw(win)\n",
    "    txt_s2b.draw(win)\n",
    "    line1_h.draw(win)\n",
    "    line1_vl.draw(win)\n",
    "    line1_vr.draw(win)\n",
    "    line2_t.draw(win)\n",
    "    line2_m.draw(win)\n",
    "    line2_b.draw(win)\n",
    "    txt_s3t.draw(win)\n",
    "    txt_s3b.draw(win)\n",
    "    line3_t.draw(win)\n",
    "    line3_m.draw(win)\n",
    "    line3_b.draw(win)\n",
    "    win.getMouse()\n",
    "    win.close()\n",
    "draw()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
