{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Enter start codon: atg\n",
      "Good Job! That's the correct start codon\n",
      "Enter a DNA triplet or stop codon: caa\n",
      "['Met', 'Gln']\n",
      "Enter a DNA triplet or stop codon: cat\n",
      "['Met', 'Gln', 'His']\n",
      "Enter a DNA triplet or stop codon: caa\n",
      "['Met', 'Gln', 'His', 'Gln']\n",
      "Enter a DNA triplet or stop codon: tag\n",
      "75\n",
      "165\n",
      "255\n",
      "345\n"
     ]
    }
   ],
   "source": [
    "import graphics\n",
    "# Nucleotide List\n",
    "DNA = [\"A\", \"C\", \"G\", \"T\"]\n",
    "\n",
    "codons = { \n",
    "    'GCT' : 'Ala',\n",
    "    'GCC' : 'Ala',\n",
    "    'GCA' : 'Ala',\n",
    "    'GCG' : 'Ala',\n",
    "    'TGT' : 'Cys',\n",
    "    'TGC' : 'Cys',\n",
    "    'GAT' : 'Asp',\n",
    "    'GAC' : 'Asp',\n",
    "    'GAA' : 'Glu',\n",
    "    'GAG' : 'Glu',\n",
    "    'TTT' : 'Phe',\n",
    "    'TTC' : 'Phe',\n",
    "    'GGT' : 'Gly',\n",
    "    'GGC' : 'Gly',\n",
    "    'GGA' : 'Gly',\n",
    "    'GGG' : 'Gly',\n",
    "    'CAT' : 'His',\n",
    "    'CAC' : 'His',\n",
    "    'ATT' : 'Ile',\n",
    "    'ATC' : 'Ile',\n",
    "    'ATA' : 'Ile',\n",
    "    'AAA' : 'Lys',\n",
    "    'AAG' : 'Lys',\n",
    "    'CTT' : 'Leu',\n",
    "    'CTC' : 'Leu',\n",
    "    'CTA' : 'Leu',\n",
    "    'CTG' : 'Leu',\n",
    "    'TTA' : 'Leu',\n",
    "    'TTG' : 'Leu',\n",
    "    'ATG' : 'Met',\n",
    "    'AAT' : 'Asn',\n",
    "    'AAC' : 'Asn',\n",
    "    'CCT' : 'Pro',\n",
    "    'CCC' : 'Pro',\n",
    "    'CCA' : 'Pro',\n",
    "    'CCG' : 'Pro',\n",
    "    'CAA' : 'Gln',\n",
    "    'CAG' : 'Gln',\n",
    "    'CGT' : 'Arg',\n",
    "    'CGC' : 'Arg',\n",
    "    'CGA' : 'Arg',\n",
    "    'CGG' : 'Arg',\n",
    "    'AGA' : 'Arg',\n",
    "    'AGG' : 'Arg',\n",
    "    'TCT' : 'Ser',\n",
    "    'TCC' : 'Ser',\n",
    "    'TCA' : 'Ser',\n",
    "    'TCG' : 'Ser',\n",
    "    'AGT' : 'Ser',\n",
    "    'AGC' : 'Ser',\n",
    "    'ACT' : 'Thr',\n",
    "    'ACC' : 'Thr',\n",
    "    'ACA' : 'Thr',\n",
    "    'ACG' : 'Thr',\n",
    "    'GTT' : 'Val',\n",
    "    'GTC' : 'Val',\n",
    "    'GTA' : 'Val',\n",
    "    'GTG' : 'Val',\n",
    "    'TGG' : 'Trp',\n",
    "    'TAT' : 'Tyr',\n",
    "    'TAC' : 'Tyr'\n",
    "           }\n",
    "\n",
    "# Stop codon list\n",
    "stop = ('TAA', 'TAG', 'TGA', 'taa', 'tag', 'tga')\n",
    "\n",
    "# Codon Dictionary\n",
    "DNA_codon_list = ['Met']\n",
    "\n",
    "def validity_check(codon):\n",
    "    codon = codon.upper()\n",
    "    if set(codon).issubset(DNA):\n",
    "        return three_base_pairs(codon)\n",
    "    else:\n",
    "        print(\"This is an invalid DNA sequence\")\n",
    "\n",
    "def three_base_pairs(codon):\n",
    "    if len(codon)/3 == 1:\n",
    "        return DNA_list(codon)\n",
    "    else:\n",
    "        print(\"This is an invalid DNA codon. Must be three bases in length.\")\n",
    "        \n",
    "def DNA_list(codon):\n",
    "    global DNA_codon_list\n",
    "    DNA_codon_list.append(codons[codon])\n",
    "    print(DNA_codon_list)\n",
    "        \n",
    "def main():\n",
    "    codon = input(\"Enter start codon: \")\n",
    "    if codon == 'ATG' or codon == 'atg':\n",
    "        print(\"Good Job! That's the correct start codon\")\n",
    "    else:\n",
    "        print(\"Sorry, that's not correct. ATG is the start codon for DNA, you may continue anyways.\")\n",
    "    codon = input(\"Enter a DNA triplet or stop codon: \")\n",
    "    while codon not in stop and len(DNA_codon_list) < 10:\n",
    "        validity_check(codon)\n",
    "        codon = input(\"Enter a DNA triplet or stop codon: \")\n",
    "\n",
    "main()\n",
    "\n",
    "def draw():\n",
    "    category = { 'Ala': '#00AAA0', 'Val': '#00AAA0', 'Cys': '#00AAA0', 'Pro': '#00AAA0', 'Gly': '#00AAA0',\n",
    "                'Leu': '#00AAA0', 'Ile': '#00AAA0', 'Met': \"#00AAA0\", 'Trp': '#00AAA0', 'Phe':'#00AAA0', \n",
    "                'Ser': '#FF7A5A', 'Thr': '#FF7A5A', 'Tyr': '#FF7A5A', 'Asn': '#FF7A5A', 'Gln':'#FF7A5A',\n",
    "                'Lys': '#001D71', 'Arg': '#001D71', 'His':'#001D71',\n",
    "                'Asp': '#853C97', 'Glu':'#853C97'\n",
    "               }\n",
    "\n",
    "    win = graphics.GraphWin(\"Primary Structure of Protein\", 1000, 500)\n",
    "    for i, amino_acids in enumerate(DNA_codon_list):\n",
    "        pos = 75 + 90 * i\n",
    "        pt = graphics.Point(x = pos, y = 250)  \n",
    "        c = graphics.Circle(pt, 45)\n",
    "        c.draw(win)\n",
    "        c.setFill(category[amino_acids])\n",
    "        text_label = graphics.Text(graphics.Point(x = pos, y = 250), amino_acids)\n",
    "        text_label.draw(win)\n",
    "        text_label.setFill(\"white\")\n",
    "    text1 = graphics.Text(graphics.Point(500, 25), \"What a pretty protein!\")\n",
    "    text2 = graphics.Text(graphics.Point(500, 50), \"Teal = Non-Polar Amino Acids\")\n",
    "    text3 = graphics.Text(graphics.Point(500, 70), \"Coral = Polar Amino Acids\")\n",
    "    text4 = graphics.Text(graphics.Point(500, 90), \"Navy = Basic Amino Acids\")\n",
    "    text5 = graphics.Text(graphics.Point(500, 110), \"Purple = Acidic Amino Acids\")\n",
    "    text6 = graphics.Text(graphics.Point(500, 450), \"Click the screen to denature your protein!\")\n",
    "    text1.draw(win)\n",
    "    text2.draw(win)\n",
    "    text3.draw(win)\n",
    "    text4.draw(win)\n",
    "    text5.draw(win)\n",
    "    text6.draw(win)\n",
    "    win.getMouse()\n",
    "    win.close()\n",
    "draw()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
